src/Bech32.php

```diff
@@ -0,0 +1,588 @@
+<?php
+
+namespace CardanoPhp\Bech32;
+
+use Exception;
+use function array_merge;
+use function array_slice;
+use function count;
+use function implode;
+use function ord;
+use function pack;
+use function strlen;
+
+/**
+ * Original inspiration for the encoding/decoding portion of this library came
+ * from https://github.com/Bit-Wasp/bech32. However, there was extra and
+ * unnecessary bloat there due to being used in other implementations e.g.
+ * Bitcoin so I have extracted only the necessary pieces here.
+ */
+class Bech32
+{
+    const GENERATOR   = [
+        0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3,
+    ];
+
+    const CHARSET     = 'qpzry9x8gf2tvdw0s3jn54khce6mua7l';
+
+    const CHARKEY_KEY = [
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 15, -1, 10, 17, 21, 20,
+        26, 30, 7, 5, -1, -1, -1, -1, -1, -1, -1, 29, -1, 24, 13, 25, 9, 8, 23,
+        -1, 18, 22, 31, 27, 19, -1, 1, 0, 3, 16, 11, 28, 12, 14, 6, 4, 2, -1,
+        -1, -1, -1, -1, -1, 29, -1, 24, 13, 25, 9, 8, 23, -1, 18, 22, 31, 27,
+        19, -1, 1, 0, 3, 16, 11, 28, 12, 14, 6, 4, 2, -1, -1, -1, -1, -1,
+    ];
+
+    /**
+     * @param string $hrp
+     * @param array  $combinedDataChars
+     *
+     * @return string
+     */
+    public static function encode(string $hrp, array $combinedDataChars): string
+    {
+        $checksum = self::createChecksum($hrp, $combinedDataChars);
+        $characters = array_merge($combinedDataChars, $checksum);
+
+        $encoded = [];
+        for ($i = 0, $n = count($characters); $i < $n; $i++) {
+            $encoded[$i] = self::CHARSET[$characters[$i]];
+        }
+
+        return "{$hrp}1" . implode('', $encoded);
+    }
+
+    /**
+     * Validates a Bech32 string and returns [$hrp, $dataChars] if
+     * the conversion was successful. An exception is thrown on invalid
+     * data.
+     *
+     * @param string $sBech - the Bech32 encoded string
+     *
+     * @return array - returns [$hrp, $dataChars]
+     * @throws Exception
+     */
+    public static function decode(string $sBech): array
```

Comment:
Minor: The `decode` does not validate that the HRP conforms to Bech32 character constraints (a-z, 0-9, case-insensitive). This could lead to issues if an HRP with invalid characters is used.

Potential validation could be done using the following regex pattern `/^[a-z0-9]+$/`

Reply:
This has been created as a new issue and marked as a "Good First Issue" for a contributor